File: examples/python/basic_lifecycle.py
#!/usr/bin/env python3
# examples/python/basic_lifecycle.py
# ============================================================================
# Module: Decision Gate Basic Lifecycle (Python)
# Description: Define a scenario, start a run, and fetch status over MCP HTTP.
# Purpose: Runnable example for SDK consumers (validated + deterministic).
# Dependencies: decision_gate SDK (local), stdlib
# ============================================================================

from __future__ import annotations

import json
import os
import sys

from decision_gate import (
    DecisionGateClient,
    validate_scenario_define_request,
    validate_scenario_start_request,
    validate_scenario_status_request,
)


def load_env_json(name: str) -> object | None:
    value = os.environ.get(name)
    if not value:
        return None
    return json.loads(value)


def default_time_after_spec(scenario_id: str, threshold: int) -> dict:
    return {
        "scenario_id": scenario_id,
        "namespace_id": 1,
        "spec_version": "1",
        "default_tenant_id": None,
        "policies": [],
        "schemas": [],
        "predicates": [
            {
                "predicate": "after",
                "query": {
                    "provider_id": "time",
                    "predicate": "after",
                    "params": {"timestamp": threshold},
                },
                "comparator": "equals",
                "expected": True,
                "policy_tags": [],
                "trust": None,
            }
        ],
        "stages": [
            {
                "stage_id": "main",
                "entry_packets": [],
                "gates": [
                    {
                        "gate_id": "gate-time",
                        "requirement": {"predicate": "after"},
                        "trust": None,
                    }
                ],
                "advance_to": {"kind": "terminal"},
                "timeout": None,
                "on_timeout": "fail",
            }
        ],
    }


def default_run_config(scenario_id: str, run_id: str) -> dict:
    return {
        "scenario_id": scenario_id,
        "run_id": run_id,
        "tenant_id": 1,
        "namespace_id": 1,
        "policy_tags": [],
        "dispatch_targets": [],
    }


def maybe_validate(enabled: bool, validator, payload: dict) -> None:
    if enabled:
        validator(payload)


def main() -> int:
    endpoint = os.environ.get("DG_ENDPOINT", "http://127.0.0.1:8080/rpc")
    token = os.environ.get("DG_TOKEN")
    validate_enabled = os.environ.get("DG_VALIDATE") == "1"

    spec = load_env_json("DG_SCENARIO_SPEC") or default_time_after_spec(
        "example-basic", 0
    )
    scenario_id = spec["scenario_id"]
    run_config = load_env_json("DG_RUN_CONFIG") or default_run_config(
        scenario_id, "run-basic-1"
    )
    started_at = load_env_json("DG_STARTED_AT") or {"kind": "logical", "value": 1}

    client = DecisionGateClient(endpoint=endpoint, auth_token=token)

    define_request = {"spec": spec}
    maybe_validate(validate_enabled, validate_scenario_define_request, define_request)
    define = client.scenario_define(define_request)

    start_request = {
        "scenario_id": scenario_id,
        "run_config": run_config,
        "started_at": started_at,
        "issue_entry_packets": False,
    }
    maybe_validate(validate_enabled, validate_scenario_start_request, start_request)
    start = client.scenario_start(start_request)

    status_request = {
        "scenario_id": scenario_id,
        "request": {
            "tenant_id": run_config["tenant_id"],
            "namespace_id": run_config["namespace_id"],
            "run_id": run_config["run_id"],
            "requested_at": {"kind": "logical", "value": 2},
            "correlation_id": None,
        },
    }
    maybe_validate(validate_enabled, validate_scenario_status_request, status_request)
    status = client.scenario_status(status_request)

    print(json.dumps({"define": define, "start": start, "status": status}))
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as exc:
        print(json.dumps({"status": "fatal_error", "error": str(exc)}))
        sys.exit(1)

File: examples/typescript/basic_lifecycle.ts
// examples/typescript/basic_lifecycle.ts
// ============================================================================
// Module: Decision Gate Basic Lifecycle (TypeScript)
// Description: Define a scenario, start a run, and fetch status over MCP HTTP.
// Purpose: Runnable example for SDK consumers (validated + deterministic).
// Dependencies: Decision Gate SDK (local), fetch API
// ============================================================================

import {
  DecisionGateClient,
  validateScenarioDefineRequestWithAjv,
  validateScenarioStartRequestWithAjv,
  validateScenarioStatusRequestWithAjv,
} from "../../sdks/typescript/src/index.ts";

function loadEnvJson(name: string): unknown | null {
  const value = process.env[name];
  if (!value) {
    return null;
  }
  return JSON.parse(value);
}

function defaultTimeAfterSpec(scenarioId: string, threshold: number): Record<string, unknown> {
  return {
    scenario_id: scenarioId,
    namespace_id: 1,
    spec_version: "1",
    default_tenant_id: null,
    policies: [],
    schemas: [],
    conditions: [
      {
        condition_id: "after",
        query: {
          provider_id: "time",
          check_id: "after",
          params: { timestamp: threshold },
        },
        comparator: "equals",
        expected: true,
        policy_tags: [],
        trust: null,
      },
    ],
    stages: [
      {
        stage_id: "main",
        entry_packets: [],
        gates: [
          {
            gate_id: "gate-time",
            requirement: { Condition: "after" },
            trust: null,
          },
        ],
        advance_to: { kind: "terminal" },
        timeout: null,
        on_timeout: "fail",
      },
    ],
  };
}

function defaultRunConfig(scenarioId: string, runId: string): Record<string, unknown> {
  return {
    scenario_id: scenarioId,
    run_id: runId,
    tenant_id: 1,
    namespace_id: 1,
    policy_tags: [],
    dispatch_targets: [],
  };
}

async function maybeValidate(enabled: boolean, validator: (payload: any) => Promise<void>, payload: any) {
  if (enabled) {
    await validator(payload);
  }
}

async function main(): Promise<number> {
  const endpoint = process.env.DG_ENDPOINT ?? "http://127.0.0.1:8080/rpc";
  const token = process.env.DG_TOKEN;
  const validateEnabled = process.env.DG_VALIDATE === "1";

  const spec = (loadEnvJson("DG_SCENARIO_SPEC") as Record<string, unknown>) ??
    defaultTimeAfterSpec("example-basic", 0);
  const scenarioId = spec.scenario_id as string;
  const runConfig = (loadEnvJson("DG_RUN_CONFIG") as Record<string, unknown>) ??
    defaultRunConfig(scenarioId, "run-basic-1");
  const startedAt = loadEnvJson("DG_STARTED_AT") ?? { kind: "logical", value: 1 };

  const client = new DecisionGateClient({
    endpoint,
    authToken: token,
  });

  const defineRequest = { spec };
  await maybeValidate(validateEnabled, validateScenarioDefineRequestWithAjv, defineRequest);
  const define = await client.scenario_define(defineRequest);

  const startRequest = {
    scenario_id: scenarioId,
    run_config: runConfig,
    started_at: startedAt,
    issue_entry_packets: false,
  };
  await maybeValidate(validateEnabled, validateScenarioStartRequestWithAjv, startRequest);
  const start = await client.scenario_start(startRequest);

  const statusRequest = {
    scenario_id: scenarioId,
    request: {
      tenant_id: runConfig.tenant_id,
      namespace_id: runConfig.namespace_id,
      run_id: runConfig.run_id,
      requested_at: { kind: "logical", value: 2 },
      correlation_id: null,
    },
  };
  await maybeValidate(validateEnabled, validateScenarioStatusRequestWithAjv, statusRequest);
  const status = await client.scenario_status(statusRequest);

  console.log(JSON.stringify({ define, start, status }));
  return 0;
}

main()
  .then((code) => process.exit(code))
  .catch((error) => {
    console.log(JSON.stringify({ status: "fatal_error", error: String(error) }));
    process.exit(1);
  });
